lib/dom_parser.rb

```diff
@@ -0,0 +1,164 @@
+#require 'pry-byebug'
+require_relative "./dom_tree.rb"
+require_relative "./html_reader.rb"
+require_relative "./html_writer.rb"
+require_relative "./node_renderer.rb"
+require_relative "./node_searcher.rb"
+require_relative "./stack.rb"
+
+class DOMParser
+  attr_reader :tree
+  attr_accessor :render, :search, :writer
+
+  def initialize(file)
+    @reader = HTMLReader.new
+    @writer = HTMLWriter.new
+    @tree = DOMTree.new
+    @render = NodeRenderer.new
+    @search = NodeSearcher.new
+    @stack = Stack.new
+    @file = file
+  end
+
+  # Looks at array from reader.html and calls correct parser for each element
+  # Returns array of hashes with each element being the data of a tree node
+  def parser(array)
+    tree_data_array = []
+    array.each do |element|
+      if @reader.is_tag?(element)
+        element_data = tag_parser(element)
+        element_data[:type] = "open"
+      elsif @reader.is_closing_tag?(element) 
+        element_data = {:type => "close"}
+      else
+        element_data = content_parser(element)
+        element_data[:type] = "content"
+      end
+      tree_data_array << element_data
+    end
+    tree_data_array
+  end
+
+  # Creates tag hash with attr-type and value pairs
+  def tag_parser(string)
+    attr_hash = {}
+
+    tag_type = string.match(/<(\w+)./)[1]
+    attr_hash[:tag_type] = tag_type
+
+    attribute_match_data = string.scan(/(\w+)\s?=\s?'([\S]*[\w\s]*)'/)
+    attribute_match_data.each do |match|
+      key = match[0]
+      values = match[1].split(" ")
+      attr_hash[key] = values
+    end
+    attr_hash
+  end
+
+  # Creates content hash for text elements
+  def content_parser(string)
+    raise(TypeError) if string.class != String
+    content_hash = {:content => string}
+  end
+
+  # Translates tag hash into useable html
+  def tag_printer(tag_hash)
```

Comment:
This function and "tree_to_string" seem unrelated to parsing, and might deserve their own class.